File: src/prototype/web/tests/unit/stamina.test.ts
// Tests for stamina consumption / recovery — docs/design/state_machines_v1.md §5.

import { describe, expect, it } from "vitest";
import {
  DEFAULT_STAMINA_CONFIG,
  applyConfirmCost,
  canStartReach,
  gripStrengthCeiling,
  updateStamina,
  type StaminaInputs,
} from "../../src/state/stamina.js";
import { initialActorState, type ActorState } from "../../src/state/game_state.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";

function gripped(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: "SLEEVE_R" as const,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function reaching(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "REACHING" as const,
    target: "SLEEVE_R" as const,
    stateEnteredMs: 0,
    reachDurationMs: 275,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function actor(over: Partial<ActorState> = {}): ActorState {
  return Object.freeze({ ...initialActorState(0), ...over });
}

function inputs(over: Partial<StaminaInputs> = {}): StaminaInputs {
  return {
    dtMs: 1000,
    actor: actor(),
    attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
    triggerL: 0,
    triggerR: 0,
    breathPressed: false,
    ...over,
  };
}

describe("stamina drain", () => {
  it("active hand (REACHING) drains at 0.02/s", () => {
    const s = updateStamina(1, inputs({
      actor: actor({ leftHand: reaching("L") }),
    }));
    expect(s).toBeCloseTo(1 - DEFAULT_STAMINA_CONFIG.handActiveDrainPerSec, 4);
  });

  it("GRIPPED with strength ≥ 0.5 drains", () => {
    const s = updateStamina(1, inputs({
      actor: actor({ leftHand: gripped("L") }),
      triggerL: 0.8,
    }));
    expect(s).toBeLessThan(1);
  });

  it("GRIPPED with low strength does not drain from the hand-active clause", () => {
    const s = updateStamina(1, inputs({
      actor: actor({ leftHand: gripped("L") }),
      triggerL: 0.3,
    }));
    expect(s).toBeCloseTo(1, 5);
  });

  it("posture_break adds a proportional drain on top of hand-active drain", () => {
    // Hand active + posture break to isolate the posture term.
    // Expected per-second drain: 0.02 (hand) + 0.05 * 0.5 (posture) = 0.045.
    const s = updateStamina(1, inputs({
      actor: actor({ leftHand: reaching("L"), postureBreak: { x: 0.5, y: 0 } }),
    }));
    expect(s).toBeCloseTo(1 - (DEFAULT_STAMINA_CONFIG.handActiveDrainPerSec + 0.025), 4);
  });
});

describe("stamina recovery", () => {
  it("BTN_BREATH + no grips + static hip → +0.1/s", () => {
    const s = updateStamina(0.5, inputs({
      breathPressed: true,
      attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
    }));
    expect(s).toBeCloseTo(0.5 + DEFAULT_STAMINA_CONFIG.breathRecoverPerSec, 4);
  });

  it("BTN_BREATH fails to recover if hip is moving", () => {
    const s = updateStamina(0.5, inputs({
      breathPressed: true,
      attackerHip: { hip_angle_target: 0, hip_push: 0.6, hip_lateral: 0 },
    }));
    // Hip is not static → idle recovery also won't fire (hip motion
    // implies motion). So stamina stays put.
    expect(s).toBeCloseTo(0.5, 4);
  });

  it("all limbs IDLE + no grips → +0.03/s", () => {
    const s = updateStamina(0.5, inputs());
    expect(s).toBeCloseTo(0.5 + DEFAULT_STAMINA_CONFIG.idleRecoverPerSec, 4);
  });
});

describe("stamina clamps and thresholds", () => {
  it("never exceeds 1", () => {
    const s = updateStamina(0.99, inputs({ breathPressed: true }));
    expect(s).toBeLessThanOrEqual(1);
  });

  it("never goes below 0", () => {
    const s = updateStamina(0.01, inputs({
      actor: actor({
        leftHand: gripped("L"),
        rightHand: gripped("R"),
        postureBreak: { x: 1, y: 0 },
      }),
      triggerL: 1,
      triggerR: 1,
    }));
    expect(s).toBeGreaterThanOrEqual(0);
  });

  it("gripStrengthCeiling caps at 0.6 when stamina < 0.2", () => {
    expect(gripStrengthCeiling(0.1)).toBe(0.6);
    expect(gripStrengthCeiling(0.5)).toBe(1);
  });

  it("canStartReach rejects when stamina < 0.05", () => {
    expect(canStartReach(0.04)).toBe(false);
    expect(canStartReach(0.05)).toBe(true);
  });

  it("applyConfirmCost deducts a flat 0.1", () => {
    expect(applyConfirmCost(0.5)).toBeCloseTo(0.4, 5);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/StaminaTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/stamina.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class StaminaTest
    {
        // --- helpers -------------------------------------------------------

        static HandFSM Gripped(HandSide side, GripZone zone = GripZone.SleeveR) => new HandFSM
        {
            Side             = side,
            State            = HandState.Gripped,
            Target           = zone,
            StateEnteredMs   = 0,
            ReachDurationMs  = 0,
            LastParriedZone  = GripZone.None,
            LastParriedAtMs  = BJJConst.SentinelTimeMs,
        };

        static HandFSM Reaching(HandSide side) => new HandFSM
        {
            Side             = side,
            State            = HandState.Reaching,
            Target           = GripZone.SleeveR,
            StateEnteredMs   = 0,
            ReachDurationMs  = 275,
            LastParriedZone  = GripZone.None,
            LastParriedAtMs  = BJJConst.SentinelTimeMs,
        };

        static ActorState Actor(
            HandFSM? leftHand  = null,
            HandFSM? rightHand = null,
            Vec2 postureBreak  = default) =>
            new ActorState
            {
                LeftHand          = leftHand  ?? GameStateOps.InitialActorState().LeftHand,
                RightHand         = rightHand ?? GameStateOps.InitialActorState().RightHand,
                LeftFoot          = GameStateOps.InitialActorState().LeftFoot,
                RightFoot         = GameStateOps.InitialActorState().RightFoot,
                PostureBreak      = postureBreak,
                Stamina           = 1f,
                ArmExtractedLeft  = false,
                ArmExtractedRight = false,
            };

        static StaminaInputs Inputs(
            float dtMs        = 1000f,
            ActorState? actor = null,
            float hipPush     = 0f,
            float hipLateral  = 0f,
            float triggerL    = 0f,
            float triggerR    = 0f,
            bool breathPressed = false) =>
            new StaminaInputs
            {
                DtMs          = dtMs,
                Actor         = actor ?? Actor(),
                AttackerHip   = new HipIntent { HipPush = hipPush, HipLateral = hipLateral },
                TriggerL      = triggerL,
                TriggerR      = triggerR,
                BreathPressed = breathPressed,
            };

        // --- drain tests -------------------------------------------------------

        [Test]
        public void ActiveHand_Reaching_Drains_At_HandActiveDrainPerSec()
        {
            float s = StaminaOps.UpdateStamina(1f, Inputs(actor: Actor(leftHand: Reaching(HandSide.L))));
            float expected = 1f - StaminaConfig.Default.HandActiveDrainPerSec;
            Assert.AreEqual(expected, s, 1e-4f);
        }

        [Test]
        public void Gripped_WithHighStrength_Drains()
        {
            float s = StaminaOps.UpdateStamina(1f,
                Inputs(actor: Actor(leftHand: Gripped(HandSide.L)), triggerL: 0.8f));
            Assert.Less(s, 1f);
        }

        [Test]
        public void Gripped_WithLowStrength_DoesNotDrain_FromHandActiveClause()
        {
            float s = StaminaOps.UpdateStamina(1f,
                Inputs(actor: Actor(leftHand: Gripped(HandSide.L)), triggerL: 0.3f));
            Assert.AreEqual(1f, s, 1e-5f);
        }

        [Test]
        public void PostureBreak_AddsDrain_ProportionalToMagnitude()
        {
            // Expected per-second: 0.02 (hand active) + 0.05 × 0.5 (posture) = 0.045
            float s = StaminaOps.UpdateStamina(1f, Inputs(
                actor: Actor(leftHand: Reaching(HandSide.L), postureBreak: new Vec2(0.5f, 0f))));
            float expected = 1f - (StaminaConfig.Default.HandActiveDrainPerSec + StaminaConfig.Default.PostureMaintainDrainPerSec * 0.5f);
            Assert.AreEqual(expected, s, 1e-4f);
        }

        // --- recovery tests -------------------------------------------------------

        [Test]
        public void BtnBreath_NoGrips_StaticHip_Recovers_AtBreathRate()
        {
            float s = StaminaOps.UpdateStamina(0.5f, Inputs(breathPressed: true));
            float expected = 0.5f + StaminaConfig.Default.BreathRecoverPerSec;
            Assert.AreEqual(expected, s, 1e-4f);
        }

        [Test]
        public void BtnBreath_MovingHip_DoesNotRecover()
        {
            float s = StaminaOps.UpdateStamina(0.5f, Inputs(breathPressed: true, hipPush: 0.6f));
            Assert.AreEqual(0.5f, s, 1e-4f);
        }

        [Test]
        public void AllLimbsIdle_NoGrips_Recovers_AtIdleRate()
        {
            float s = StaminaOps.UpdateStamina(0.5f, Inputs());
            float expected = 0.5f + StaminaConfig.Default.IdleRecoverPerSec;
            Assert.AreEqual(expected, s, 1e-4f);
        }

        // --- clamps / thresholds -------------------------------------------------------

        [Test]
        public void Stamina_NeverExceeds_One()
        {
            float s = StaminaOps.UpdateStamina(0.99f, Inputs(breathPressed: true));
            Assert.LessOrEqual(s, 1f);
        }

        [Test]
        public void Stamina_NeverGoesBelow_Zero()
        {
            float s = StaminaOps.UpdateStamina(0.01f, Inputs(
                actor: Actor(
                    leftHand:  Gripped(HandSide.L),
                    rightHand: Gripped(HandSide.R),
                    postureBreak: new Vec2(1f, 0f)),
                triggerL: 1f, triggerR: 1f));
            Assert.GreaterOrEqual(s, 0f);
        }

        [Test]
        public void GripStrengthCeiling_CapAt0_6_WhenStaminaBelow0_2()
        {
            Assert.AreEqual(0.6f, StaminaOps.GripStrengthCeiling(0.1f));
            Assert.AreEqual(1f,   StaminaOps.GripStrengthCeiling(0.5f));
        }

        [Test]
        public void CanStartReach_Rejects_WhenStaminaBelow0_05()
        {
            Assert.IsFalse(StaminaOps.CanStartReach(0.04f));
            Assert.IsTrue( StaminaOps.CanStartReach(0.05f));
        }

        [Test]
        public void ApplyConfirmCost_Deducts_Flat0_1()
        {
            Assert.AreEqual(0.4f, StaminaOps.ApplyConfirmCost(0.5f), 1e-5f);
        }
    }
}
